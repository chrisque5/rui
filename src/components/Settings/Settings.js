import PropTypes from 'prop-types';
import {
  Modal, Button, Switch, Select, Tabs,
} from 'antd';
import { settingsProperties } from '../../utils/constants';

const Settings = ({
  visible, onSave, onBack, settings, userDefaults, getUserDefault,
  getDefaultSetting, onChanged, onUserDefaultChanged, firmSettingOnCallback,
}) => {
  // if displayClientFavourites determine whether to disable it or not
  // else return default idasbled flag
  const isControlDisabled = (disabled, key, firmSettingOnValue) => {
    if (key === 'displayClientFavourites') {
      // if the display clients setting has been updated before save use that setting
      // else use the displayClients value
      // else use the default disabled value
      return firmSettingOnValue !== undefined ? firmSettingOnValue : getDefaultSetting('displayClients') || disabled;
    }

    return disabled;
  };

  // callback to update the firm setting on value
  const firmSettingOn = firmSettingOnCallback();

  const renderSelect = (item, options) => (
    <Select
      style={{ minWidth: 60 }}
      onChange={(value, event) => onUserDefaultChanged(value, event, item)}
      data-testid={`select-${item}`}
      defaultValue={getUserDefault(item)}
    >
      { Object.entries(options).map(
        ([, { key, text }]) => (
          <Select.Option key={key} value={key}>{text}</Select.Option>
        ),
      )}
    </Select>
  );

  /**
   * Render a specific control type
   *
   * @param {string} type the contol type
   * @param {string} key the setting key
   */
  const renderControl = (key, type, disabled) => {
    const defaultSetting = getDefaultSetting(key);

    if (type === 'switch') {
      return (
        <Switch
          disabled={disabled}
          data-testid={`chkSwitch-${key}`}
          checkedChildren="on"
          unCheckedChildren="off"
          defaultChecked={defaultSetting}
          onChange={(value, event) => onChanged(value, event, key)}
        />
      );
    }

    if (type === 'select') {
      return renderSelect(key, userDefaults[key].options);
    }

    return (<span>{defaultSetting}</span>);
  };

  /**
   * Render all the settings. Loop through all the entries in the
   * settingsProperties map and create a setting
   */
  const renderSettings = () => {
    if (settings && settingsProperties) {
      return (
        <Tabs tabPosition="left">
          {
            settingsProperties.map(({ tabKey, tabLabel, items }) => (
              <Tabs.TabPane tab={tabLabel} key={tabKey} className="setting-section">
                {Object.entries(items).map(
                  ([key, { disabled = false, label, type }]) => {
                    const isDisabled = isControlDisabled(disabled, key, firmSettingOn);
                    return (
                      <div className="setting" key={key}>
                        <span>{label}</span>
                        { renderControl(key, type, isDisabled)}
                      </div>
                    );
                  },
                )}
              </Tabs.TabPane>
            ))
          }
        </Tabs>
      );
    }

    return (
      <span>Failed to retrieve favorites from server, please reload the page.</span>
    );
  };

  return (
    <Modal
      className="settings-modal"
      title="User settings"
      visible={visible}
      onOk={onSave}
      onCancel={onBack}
      bodyStyle={{ padding: '10px', minHeight: 150 }}
      footer={[
        <Button data-testid="btnBack" key="back" onClick={onBack}>Back</Button>,
        <Button data-testid="btnSave" key="submit" type="primary" onClick={onSave}>Save</Button>,
      ]}
      destroyOnClose
    >
      {renderSettings()}
    </Modal>
  );
};

Settings.defaultProps = {
  firmSettingOnCallback: null,
};

Settings.propTypes = {
  onSave: PropTypes.func.isRequired,
  onBack: PropTypes.func.isRequired,
  onChanged: PropTypes.func.isRequired,
  onUserDefaultChanged: PropTypes.func.isRequired,
  settings: PropTypes.shape().isRequired,
  userDefaults: PropTypes.shape().isRequired,
  getDefaultSetting: PropTypes.func.isRequired,
  getUserDefault: PropTypes.func.isRequired,
  visible: PropTypes.bool.isRequired,
  firmSettingOnCallback: PropTypes.func,
};

export default Settings;
