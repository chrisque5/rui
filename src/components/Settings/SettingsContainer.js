import { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import Settings from './Settings';
import { changeCounterPartySelection, toggleSettingsModal } from '../../actions/uiActions';
import { changePreference } from '../../actions/userActions';
import {
  getIsSettingsModalVisible, getUserPreferenceSettings, getUserPermissions, getUserPreferencesDefaults,
} from '../../utils/selectors';
import { dealTypes, navMenuItems } from '../../utils/constants';
import { isObjectEmpty } from '../../utils/helper';

const SettingsContainer = (props) => {
  const dispatch = useDispatch();
  const [changedSettings, setChangedSettings] = useState({});
  const [changedDefaults, setChangedDefaults] = useState({});

  const settings = useSelector(getUserPreferenceSettings);
  const visible = useSelector(getIsSettingsModalVisible);
  const permissions = useSelector(getUserPermissions);
  const defaults = useSelector(getUserPreferencesDefaults);

  const {
    validForDeskEdit, validForFwd, validForNdf, validForSpt, validForBlotterView,
  } = permissions;
  const { dealType } = props;

  // when the settings change notify the child
  const firmSettingOnCallback = useCallback(() => changedSettings.displayClients, [changedSettings]);

  const onBack = () => {
    setChangedDefaults({});
    setChangedSettings({});
    dispatch(toggleSettingsModal());
  };

  const onSave = () => {
    const newSettings = {
      ...((settings && !isObjectEmpty(changedSettings)) && { settings: changedSettings }),
      ...((defaults && !isObjectEmpty(changedDefaults)) && { defaults: changedDefaults }),
    };

    if (!isObjectEmpty(newSettings)) {
      dispatch(changePreference(newSettings));

      if (!isObjectEmpty(changedSettings) && dealType === dealTypes.FWD) {
        if (changedSettings.lrMode === true) {
          dispatch(changeCounterPartySelection('seller'));
        } else if (changedSettings.lrMode === false) {
          dispatch(changeCounterPartySelection('buyer'));
        }
      }
    }
    onBack();
  };

  const onChanged = (value, event, propertyName) => {
    const newSettings = {
      ...changedSettings,
      [propertyName]: value,
    };

    setChangedSettings(newSettings);
  };

  const onUserDefaultChanged = (value, event, propertyName) => {
    const newDefaults = {
      ...changedDefaults,
      [propertyName]: value,
    };

    setChangedDefaults(newDefaults);
  };

  const {
    ADMIN, BLOTTER, FWD, NDF, SPT,
  } = navMenuItems;

  const userDefaults = {
    defaultEntryPage: {
      options: {
        ...(validForBlotterView && { BLOTTER }),
        ...(validForNdf && { NDF }),
        ...(validForFwd && { FWD }),
        ...(validForSpt && { SPT }),
        ...(validForDeskEdit && { ADMIN }),
      },
    },
  };

  const getDefaultSetting = (property) => (settings ? settings[property] : null);
  const getUserDefault = (property) => (defaults && userDefaults[property].options[defaults[property]] ? defaults[property] : null);

  return (
    <Settings
      visible={visible}
      onSave={onSave}
      onBack={onBack}
      settings={settings}
      userDefaults={userDefaults}
      getUserDefault={getUserDefault}
      getDefaultSetting={getDefaultSetting}
      onChanged={onChanged}
      onUserDefaultChanged={onUserDefaultChanged}
      firmSettingOnCallback={firmSettingOnCallback}
    />
  );
};

SettingsContainer.propTypes = {
  dealType: PropTypes.string.isRequired,
};

export default SettingsContainer;
