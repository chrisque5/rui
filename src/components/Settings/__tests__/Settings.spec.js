import {
  render, fireEvent, waitFor,
} from 'test-utils/testUtils';
import Settings from '../Settings';

const getDefaultSetting = jest.fn();
const getUserDefault = jest.fn();

const onBack = jest.fn();
const onSave = jest.fn();
const onChanged = jest.fn();
const onUserDefaultChanged = jest.fn();
const firmSettingOnCallback = jest.fn();

const defaultProps = {
  visible: false,
  onSave,
  onBack,
  settings: {},
  userDefaults: {
    defaultEntryPage: {
      options: {
        ADMIN: {
          key: 'ADMIN', text: 'Admin', to: '/admin', documentTitle: 'Admin',
        },
        BLOTTER: {
          key: 'BLOTTER', target: '_blank', text: 'Blotter', to: '/blotter', documentTitle: 'Blotter',
        },
      },
    },
  },
  getDefaultSetting,
  getUserDefault,
  onChanged,
  onUserDefaultChanged,
  firmSettingOnCallback,
};

afterEach(() => {
  jest.clearAllMocks();
});

const renderUI = (props = {}) => {
  const rtlProps = render(<Settings {...defaultProps} {...props} />);

  return {
    ...rtlProps,
    rerender: (newProps) => rtlProps.rerender(<Settings {...defaultProps} {...props} {...newProps} />),
  };
};

test('<Settings /> renders snapshot as expected', () => {
  const { container } = renderUI();
  expect(container).toMatchSnapshot();
});

test('<Settings /> modal is displayed when the visible flag is set to true', () => {
  const { queryByText, rerender } = renderUI();
  let modal = queryByText('User settings');
  expect(modal).toBeNull();
  rerender({ visible: true });
  modal = queryByText('User settings');
  expect(modal).not.toBeNull();
});

test('<Settings /> Back button calls onBack function', () => {
  const { getByText } = renderUI({ visible: true });
  fireEvent.click(getByText(/Back/i));
  expect(onBack).toHaveBeenCalledTimes(1);
});

test('<Settings /> Save button calls onSave function', () => {
  const { getByText } = renderUI({ visible: true });
  fireEvent.click(getByText(/Save/i));
  expect(onSave).toHaveBeenCalledTimes(1);
});

describe('<Settings /> Tailor Rates Feed Section', () => {
  ['ratesFeed'].forEach((setting) => {
    test(`${setting} button click switches the button off and calls onChange function with false when default setting is true`,
      () => {
        getDefaultSetting.mockReturnValue(true);
        const { getByText, getByTestId } = renderUI({ visible: true });
        fireEvent.click(getByText(/Tailor Rates Feed/i));

        const button = getByTestId(`chkSwitch-${setting}`);

        expect(button.textContent).toBe('on');
        fireEvent.click(getByTestId(`chkSwitch-${setting}`));
        expect(onChanged).toHaveBeenCalledTimes(1);
        expect(onChanged).toHaveBeenCalledWith(false, expect.anything(), setting);
        expect(button.textContent).toBe('off');
      });

    test(`${setting} button click switches the button on and calls onChange function with true when default setting is false`,
      () => {
        getDefaultSetting.mockReturnValue(false);
        const { getByText, getByTestId } = renderUI({ visible: true });
        fireEvent.click(getByText(/Tailor Rates Feed/i));

        const button = getByTestId(`chkSwitch-${setting}`);

        expect(button.textContent).toBe('off');
        fireEvent.click(getByTestId(`chkSwitch-${setting}`));
        expect(onChanged).toHaveBeenCalledTimes(1);
        expect(onChanged).toHaveBeenCalledWith(true, expect.anything(), setting);
        expect(button.textContent).toBe('on');
      });
  });
});

describe('<Settings /> Client Manager Section', () => {
  [
    'displayClients', 'displayCurrencyFavourites',
    'displayTermFavourites', 'displayExecutionVenueFavourites',
  ].forEach((setting) => {
    test(`${setting} button click switches the button off and calls onChange function with false when default setting is true`,
      () => {
        getDefaultSetting.mockReturnValue(true);
        const { getByText, getByTestId } = renderUI({ visible: true });
        fireEvent.click(getByText(/Client Manager/i));

        const button = getByTestId(`chkSwitch-${setting}`);

        expect(button.textContent).toBe('on');
        fireEvent.click(getByTestId(`chkSwitch-${setting}`));
        expect(onChanged).toHaveBeenCalledTimes(1);
        expect(onChanged).toHaveBeenCalledWith(false, expect.anything(), setting);
        expect(button.textContent).toBe('off');
      });

    test(`${setting} button click switches the button on and calls onChange function with true when default setting is false`,
      () => {
        getDefaultSetting.mockReturnValue(false);
        const { getByText, getByTestId } = renderUI({ visible: true });
        fireEvent.click(getByText(/Client Manager/i));

        const button = getByTestId(`chkSwitch-${setting}`);

        expect(button.textContent).toBe('off');
        fireEvent.click(getByTestId(`chkSwitch-${setting}`));
        expect(onChanged).toHaveBeenCalledTimes(1);
        expect(onChanged).toHaveBeenCalledWith(true, expect.anything(), setting);
        expect(button.textContent).toBe('on');
      });
  });
});

describe('<Settings /> General Section', () => {
  ['displayExecutionVenueColours', 'displayClsDefaults'].forEach((setting) => {
    test(`${setting} button click switches the button off and calls onChange function with false when default setting is true`,
      () => {
        getDefaultSetting.mockReturnValue(true);
        const { getByText, getByTestId } = renderUI({ visible: true });
        fireEvent.click(getByText(/General/i));

        const button = getByTestId(`chkSwitch-${setting}`);

        expect(button.textContent).toBe('on');
        fireEvent.click(getByTestId(`chkSwitch-${setting}`));
        expect(onChanged).toHaveBeenCalledTimes(1);
        expect(onChanged).toHaveBeenCalledWith(false, expect.anything(), setting);
        expect(button.textContent).toBe('off');
      });

    test(`${setting} button click switches the button on and calls onChange function with true when default setting is false`,
      () => {
        getDefaultSetting.mockReturnValue(false);
        const { getByText, getByTestId } = renderUI({ visible: true });
        fireEvent.click(getByText(/General/i));

        const button = getByTestId(`chkSwitch-${setting}`);

        expect(button.textContent).toBe('off');
        fireEvent.click(getByTestId(`chkSwitch-${setting}`));
        expect(onChanged).toHaveBeenCalledTimes(1);
        expect(onChanged).toHaveBeenCalledWith(true, expect.anything(), setting);
        expect(button.textContent).toBe('on');
      });
  });

  test('Default Entry Page verify change event', async () => {
    const { getByTestId, getByText } = renderUI({ visible: true });
    fireEvent.click(getByText(/General/i));
    const dropdown = await waitFor(() => getByTestId('select-defaultEntryPage').firstElementChild);
    fireEvent.mouseDown(dropdown);

    const selectedElem = await waitFor(() => getByText(/Admin/));
    fireEvent.click(selectedElem);

    expect(onUserDefaultChanged).toHaveBeenCalledTimes(1);
  });
});

describe('<Settings /> FWD Trade Capture Section', () => {
  ['lrMode'].forEach((setting) => {
    test(`${setting} button click switches the button off and calls onChange function with false when default setting is true`,
      () => {
        getDefaultSetting.mockReturnValue(true);
        const { getByText, getByTestId } = renderUI({ visible: true });
        fireEvent.click(getByText(/FWD Trade Capture/i));

        const button = getByTestId(`chkSwitch-${setting}`);

        expect(button.textContent).toBe('on');
        fireEvent.click(getByTestId(`chkSwitch-${setting}`));
        expect(onChanged).toHaveBeenCalledTimes(1);
        expect(onChanged).toHaveBeenCalledWith(false, expect.anything(), setting);
        expect(button.textContent).toBe('off');
      });

    test(`${setting} button click switches the button on and calls onChange function with true when default setting is false`,
      () => {
        getDefaultSetting.mockReturnValue(false);
        const { getByText, getByTestId } = renderUI({ visible: true });
        fireEvent.click(getByText(/FWD Trade Capture/i));

        const button = getByTestId(`chkSwitch-${setting}`);

        expect(button.textContent).toBe('off');
        fireEvent.click(getByTestId(`chkSwitch-${setting}`));
        expect(onChanged).toHaveBeenCalledTimes(1);
        expect(onChanged).toHaveBeenCalledWith(true, expect.anything(), setting);
        expect(button.textContent).toBe('on');
      });
  });
});
