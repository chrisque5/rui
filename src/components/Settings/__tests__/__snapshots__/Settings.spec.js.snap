// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> renders snapshot as expected 1`] = `<div />`;
